rockcraft: init at 1.2.2

jnsgruk opened the pull request:
## Description of changes

This PR introduces the [`rockcraft`](https://canonical-rockcraft.readthedocs-hosted.com/en/latest/) application, which is a tool for building OCI container images.

Before the `rockcraft` package itself is added, there are some more commits adding the required Python libraries. I will follow this PR with more for `snapcraft` and `charmcraft` - hence why I have introduced `_0` and `_1` variants of packages, as those rely on different versions at this time.

I have been maintaining these in my own flake, and they've been stable for some time: https://github.com/jnsgruk/crafts-flake.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

jnsgruk committed:
rockcraft: init at 1.2.2

jnsgruk committed:
python3Packages.spdx: init at 2.5

jnsgruk committed:
python3Packages.spdx-lookup: init at 0.3.2

jnsgruk committed:
python3Packages.snap-helpers: init at 0.4.2

jnsgruk committed:
python3Packages.pydantic-yaml-0: init at 0.11.2

This introduces pydantic 0.x, which is common among packages which
have yet to migrate to pydantic 2.x.

jnsgruk committed:
python3Packages.craft-providers: init at 1.23.0

jnsgruk committed:
python3Packages.craft-parts: init at 1.26.2

jnsgruk committed:
python3Packages.craft-grammar: init at 1.1.2

jnsgruk committed:
python3Packages.craft-cli: init at 2.5.1

jnsgruk committed:
python3Packages.craft-archives: init at 1.1.3

jnsgruk commented:
Result of `nixpkgs-review pr 295646` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>24 packages built:</summary>
  <ul>
    <li>python311Packages.craft-application_1</li>
    <li>python311Packages.craft-application_1.dist</li>
    <li>python311Packages.craft-archives</li>
    <li>python311Packages.craft-archives.dist</li>
    <li>python311Packages.craft-cli</li>
    <li>python311Packages.craft-cli.dist</li>
    <li>python311Packages.craft-grammar</li>
    <li>python311Packages.craft-grammar.dist</li>
    <li>python311Packages.craft-parts</li>
    <li>python311Packages.craft-parts.dist</li>
    <li>python311Packages.craft-providers</li>
    <li>python311Packages.craft-providers.dist</li>
    <li>python311Packages.pydantic-yaml_0</li>
    <li>python311Packages.pydantic-yaml_0.dist</li>
    <li>python311Packages.snap-helpers</li>
    <li>python311Packages.snap-helpers.dist</li>
    <li>python311Packages.spdx</li>
    <li>python311Packages.spdx-lookup</li>
    <li>python311Packages.spdx-lookup.dist</li>
    <li>python311Packages.spdx.dist</li>
    <li>python312Packages.spdx</li>
    <li>python312Packages.spdx.dist</li>
    <li>rockcraft</li>
    <li>rockcraft.dist</li>
  </ul>
</details>

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/1016/104

Lurkki14 commented on pkgs/by-name/ro/rockcraft/package.nix:
```diff
@@ -0,0 +1,54 @@
+    owner = "canonical";
+    repo = "rockcraft";
+    rev = "refs/tags/${version}";
+    sha256 = "sha256-f6qLvk5cebHjZx51nJYYAwILGCTyj6FKlb3HGt6iz4M=";
```
```suggestion
    hash = "sha256-f6qLvk5cebHjZx51nJYYAwILGCTyj6FKlb3HGt6iz4M=";
```

Lurkki14 commented on pkgs/development/python-modules/craft-application_1/default.nix:
```diff
@@ -0,0 +1,51 @@
+    owner = "canonical";
+    repo = "craft-application";
+    rev = "refs/tags/${version}";
+    sha256 = "sha256-CXZEWVoE66dlQJp4G8tinufjyaDJaH1Muxz/qd/81oA=";
```
```suggestion
    hash = "sha256-CXZEWVoE66dlQJp4G8tinufjyaDJaH1Muxz/qd/81oA=";
```

Lurkki14 commented on pkgs/development/python-modules/craft-application_1/default.nix:
```diff
@@ -0,0 +1,51 @@
+    setuptools-scm
+  ];
+
+  env.SETUPTOOLS_SCM_PRETEND_VERSION = version;
```
```suggestion
```
This shouldn't be needed anymore, see https://github.com/NixOS/nixpkgs/commit/8056f9250ce8b7f9bdf0ccdc6c7a6808ae8f1be9

Lurkki14 commented on pkgs/by-name/ro/rockcraft/package.nix:
```diff
@@ -0,0 +1,54 @@
+    description = "Create OCI images using the language from Snapcraft and Charmcraft";
+    homepage = "https://github.com/canonical/rockcraft";
+    changelog = "https://github.com/canonical/rockcraft/releases/tag/${version}";
+    license = lib.licenses.lgpl3;
```
These should be the more specific `lgpl3Plus` or `lgpl3Only`

Lurkki14 commented on pkgs/development/python-modules/craft-archives/default.nix:
```diff
@@ -0,0 +1,51 @@
+    owner = "canonical";
+    repo = "craft-archives";
+    rev = "refs/tags/${version}";
+    sha256 = "sha256-ZUqMjbOsHwzZyn0NsSTlZTljzagYEirWKEGatXVL43g=";
```
```suggestion
    hash = "sha256-ZUqMjbOsHwzZyn0NsSTlZTljzagYEirWKEGatXVL43g=";
```

Lurkki14 commented on pkgs/development/python-modules/craft-archives/default.nix:
```diff
@@ -0,0 +1,51 @@
+    tabulate
+  ];
+
+  env.SETUPTOOLS_SCM_PRETEND_VERSION = version;
```
```suggestion
```

Lurkki14 commented on pkgs/development/python-modules/craft-cli/default.nix:
```diff
@@ -0,0 +1,48 @@
+    owner = "canonical";
+    repo = "craft-cli";
+    rev = "refs/tags/${version}";
+    sha256 = "sha256-yEKF04OPu4paRrghAP78r9hu6cqkUy6z/V7cHNys82I=";
```
```suggestion
    hash = "sha256-yEKF04OPu4paRrghAP78r9hu6cqkUy6z/V7cHNys82I=";
```

Lurkki14 commented on pkgs/development/python-modules/craft-cli/default.nix:
```diff
@@ -0,0 +1,48 @@
+    setuptools-scm
+  ];
+
+  SETUPTOOLS_SCM_PRETEND_VERSION = version;
```
```suggestion
```

Lurkki14 commented on pkgs/development/python-modules/craft-grammar/default.nix:
```diff
@@ -0,0 +1,38 @@
+    owner = "canonical";
+    repo = "craft-grammar";
+    rev = "refs/tags/${version}";
+    sha256 = "sha256-23KLIO2yHXGe/zb3B8LirJsh+LY9z0c5ZGtF392Kszo";
```
```suggestion
    hash = "sha256-23KLIO2yHXGe/zb3B8LirJsh+LY9z0c5ZGtF392Kszo";
```

Lurkki14 commented on pkgs/development/python-modules/craft-grammar/default.nix:
```diff
@@ -0,0 +1,38 @@
+    setuptools
+  ];
+
+  env.SETUPTOOLS_SCM_PRETEND_VERSION = version;
```
```suggestion
```

Lurkki14 commented on pkgs/development/python-modules/craft-parts/default.nix:
```diff
@@ -0,0 +1,53 @@
+    owner = "canonical";
+    repo = "craft-parts";
+    rev = "refs/tags/${version}";
+    sha256 = "sha256-wHv0JWffS916RK4Kgk+FuRthx+ajh0Ka4DBwGrLdUBs=";
```
```suggestion
    hash = "sha256-wHv0JWffS916RK4Kgk+FuRthx+ajh0Ka4DBwGrLdUBs=";
```

Lurkki14 commented on pkgs/development/python-modules/craft-providers/default.nix:
```diff
@@ -0,0 +1,59 @@
+    owner = "canonical";
+    repo = "craft-providers";
+    rev = "refs/tags/${version}";
+    sha256 = "sha256-9ZoNgpuGytwozRsw0wnS3d2UBOIsh3VI/uzB1RD2Zac=";
```
```suggestion
    hash = "sha256-9ZoNgpuGytwozRsw0wnS3d2UBOIsh3VI/uzB1RD2Zac=";
```

Lurkki14 commented on pkgs/development/python-modules/craft-providers/default.nix:
```diff
@@ -0,0 +1,59 @@
+    urllib3
+  ];
+
+  SETUPTOOLS_SCM_PRETEND_VERSION = version;
```
```suggestion
```

Lurkki14 commented on pkgs/development/python-modules/pydantic-yaml_0/default.nix:
```diff
@@ -0,0 +1,52 @@
+    owner = "NowanIlfideme";
+    repo = "pydantic-yaml";
+    rev = "refs/tags/v${version}";
+    sha256 = "sha256-AeUyVav0/k4Fz69Qizn4hcJKoi/CDR9eUan/nJhWsDY=";
```
```suggestion
    hash = "sha256-AeUyVav0/k4Fz69Qizn4hcJKoi/CDR9eUan/nJhWsDY=";
```

Lurkki14 commented on pkgs/development/python-modules/pydantic-yaml_0/default.nix:
```diff
@@ -0,0 +1,52 @@
+    setuptools-scm
+  ];
+
+  env.SETUPTOOLS_SCM_PRETEND_VERSION = version;
```
```suggestion
```

Lurkki14 commented on pkgs/development/python-modules/snap-helpers/default.nix:
```diff
@@ -0,0 +1,39 @@
+    owner = "albertodonato";
+    repo = "snap-helpers";
+    rev = "refs/tags/${version}";
+    sha256 = "sha256-7JBvrD4WNOcFSVx3xauk4JgiVGKWuUEdUMRlH7mudE4=";
```
```suggestion
    hash = "sha256-7JBvrD4WNOcFSVx3xauk4JgiVGKWuUEdUMRlH7mudE4=";
```

Lurkki14 commented on pkgs/development/python-modules/snap-helpers/default.nix:
```diff
@@ -0,0 +1,39 @@
+    setuptools
+  ];
+
+  SETUPTOOLS_SCM_PRETEND_VERSION = version;
```
```suggestion
```

jnsgruk commented:
@Lurkki14 nice, thank you! I've made those changes and it all still seems to build just fine! Thanks for the tip on the new setuptools-scm hook :slightly_smiling_face:

jnsgruk commented:
@RaitoBezarius @SuperSandro2000 could I get a little help here please? :slightly_smiling_face:

SuperSandro2000 commented on pkgs/development/python-modules/spdx-lookup/default.nix:
```diff
@@ -0,0 +1,33 @@
+    hash = "sha256-jtzhqRAj1BWdU8AuR7Gr343mL5alLXhi+SyCkCI5AAU=";
+  };
+
+  propagatedBuildInputs = with python3.pkgs; [
```
We cannot use python3 inside of pythonPackages. Please use the packages directly as inputs

SuperSandro2000 commented on pkgs/development/python-modules/snap-helpers/default.nix:
```diff
@@ -0,0 +1,37 @@
+    hash = "sha256-7JBvrD4WNOcFSVx3xauk4JgiVGKWuUEdUMRlH7mudE4=";
+  };
+
+  propagatedBuildInputs = with python3.pkgs; [
```
same here

SuperSandro2000 commented on pkgs/development/python-modules/snap-helpers/default.nix:
```diff
@@ -0,0 +1,37 @@
+  pname = "snap-helpers";
+  version = "0.4.2";
+
+  format = "pyproject";
```
This should be ``pyproject = true``

SuperSandro2000 commented on pkgs/development/python-modules/snap-helpers/default.nix:
```diff
@@ -0,0 +1,37 @@
+    pyyaml
+    setuptools
+  ];
+
```
Please execute tests https://github.com/albertodonato/snap-helpers/tree/main/tests

SuperSandro2000 commented on pkgs/development/python-modules/snap-helpers/default.nix:
```diff
@@ -0,0 +1,37 @@
+  };
+
+  propagatedBuildInputs = with python3.pkgs; [
+    packaging
```
Not sure if packaging is required at all

SuperSandro2000 commented on pkgs/development/python-modules/craft-providers/default.nix:
```diff
@@ -0,0 +1,57 @@
+
+    # The urllib3 incompat: https://github.com/msabramo/requests-unixsocket/pull/69
+    # This is already patched in nixpkgs.
+    substituteInPlace pyproject.toml \
+      --replace-fail "setuptools==67.8.0" "setuptools" \
+      --replace-fail "urllib3<2" "urllib3"
```
Can we use relaxDepsHook here?

jnsgruk replied:
I can't seem to make that work with pyproject.toml projects - tried a bunch of different configs with it in this project and couldn't make it happen.

SuperSandro2000 replied:
That's fine, we don't need to use it.

SuperSandro2000 reviewed (changes requested):
Please add pythonImportsCheck to all packages. Also if there are tests, we should execute them. I didn't check all packages for them.

I also stopped commenting the same thing on the second occurrence or so.

jnsgruk committed:
python3Packages.craft-application-1: init at 1.2.1

More craft applications will follow this initial commit. At present,
some are reliant on craft-application 1.x, and others on 2.x.

jnsgruk commented:
@SuperSandro2000 thank you! I've addressed your comments all bar one on `pythonRelaxDepsHook` which didn't seem to work for these cases.

drupol commented on pkgs/by-name/ro/rockcraft/package.nix:
```diff
@@ -0,0 +1,57 @@
+
+  passthru.updateScript = nix-update-script { };
+
+  meta = {
```
Missing `meta.mainProgram`.

drupol replied:
/me needs some sleep.

drupol reviewed (approved):
LGTM now! thanks for making the changes.

SuperSandro2000 commented on pkgs/development/python-modules/craft-parts/default.nix:
```diff
@@ -0,0 +1,122 @@
+  ];
+
+  postPatch = ''
+    # substituteInPlace craft_parts/__init__.py \
+    #   --replace-fail "dev" "${version}"
+
```
Do we still need that?

jnsgruk commented:
@SuperSandro2000 can I beg another review please? Thanks in advance 🙏

jnsgruk commented:
Thank you! 🤩

dotlambda commented on pkgs/top-level/python-packages.nix:
```diff
@@ -10529,6 +10541,8 @@ self: super: with self; {
 
   pydantic-settings = callPackage ../development/python-modules/pydantic-settings { };
 
+  pydantic-yaml-0 = callPackage ../development/python-modules/pydantic-yaml-0 { };
```
We only allow multiple versions in `python-packages.nix` in very rare cases. Why was this done?
Anyway, it should have been
```suggestion
pydantic-yaml_0 = callPackage ../development/python-modules/pydantic-yaml/0.nix { };
```

dotlambda replied:
I'm rectifying this in https://github.com/NixOS/nixpkgs/pull/329667.

jnsgruk replied:
Hey! Nice, thank you. I'll take a look!

dotlambda commented on pkgs/development/python-modules/pydantic-yaml_0/default.nix:
```diff
@@ -0,0 +1,50 @@
+}:
+
+buildPythonPackage rec {
+  pname = "pydantic_yaml_0";
```
But https://pypi.org/project/pydantic-yaml-0/ doesn't exist.
